packages/go_router/example/lib/stateful_shell_route.dart

```diff
@@ -0,0 +1,550 @@
+                  /// builder. Custom implementation can access the branch
+                  /// navigators via the StatefulShellRouteState
+                  /// (see TabbedRootScreen for details).
+                  return const TabbedRootScreen();
+                },
+              ),
+            ],
+          ),
+        ],
+        builder: (BuildContext context, GoRouterState state, Widget child) {
+          return ScaffoldWithNavBar(body: child);
+        },
+
+        /// It's possible to customize the container for the branch navigators
+        /// even further, to for instance setup custom animations. The code
+        /// below is an example of such a customization (see
+        /// _AnimatedRouteBranchContainer). Note that in this case, the Widget
+        /// provided in the child parameter should not be added to the widget
+        /// tree. Instead, access the child widgets of each branch directly
+        /// (see StatefulShellRouteState.children) to implement a custom layout
+        /// and container for the navigators.
+        // builder: (BuildContext context, GoRouterState state, Widget child) {
+        //   return ScaffoldWithNavBar(
+        //     body: _AnimatedRouteBranchContainer(),
+        //   );
+        // },
+
+        /// If it's necessary to customize the Page for StatefulShellRoute,
+        /// provide a pageBuilder function in addition to the builder, for example:
+        // pageBuilder:
+        //     (BuildContext context, GoRouterState state, Widget statefulShell) {
+        //   return NoTransitionPage<dynamic>(child: statefulShell);
+        // },
+      ),
+    ],
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      routerConfig: _router,
+    );
+  }
+}
+
+/// Builds the "shell" for the app by building a Scaffold with a
+/// BottomNavigationBar, where [child] is placed in the body of the Scaffold.
+class ScaffoldWithNavBar extends StatelessWidget {
+  /// Constructs an [ScaffoldWithNavBar].
+  const ScaffoldWithNavBar({
+    required this.body,
+    Key? key,
+  }) : super(key: key ?? const ValueKey<String>('ScaffoldWithNavBar'));
+
+  /// Body, i.e. the index stack
+  final Widget body;
+
+  @override
+  Widget build(BuildContext context) {
+    final StatefulShellRouteState shellState =
+        StatefulShellRouteState.of(context);
+
+    return Scaffold(
+      body: body,
+      bottomNavigationBar: BottomNavigationBar(
+        items: const <BottomNavigationBarItem>[
+          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Section A'),
+          BottomNavigationBarItem(icon: Icon(Icons.work), label: 'Section B'),
+          BottomNavigationBarItem(icon: Icon(Icons.tab), label: 'Section C'),
+        ],
+        currentIndex: shellState.currentIndex,
+        onTap: (int tappedIndex) => shellState.goBranch(index: tappedIndex),
```

Comment:
Well, this is a little bit different than simple location based navigation, since we’re actually rather trying to restore a complete navigation stack here, preferably with as little rebuilding as possible. So it wouldn’t work with just the location string - you need the whole `RouteMatchList`. In the other branch, I added support for using a `GoRouterState` to navigate (`GoRouter.goState`), made possible by keeping a reference to the `RouteMatchList` (actually `UnmodifiableRouteMatchList`) in the state. And it might be possible to build an even better API around this, but that might involve creating a method on `GoRouter` called perhaps `goBranch`, accepting a `StatefulShellBranchState` as argument.

Reply:
Ah so this is in case where there are imperative push to the stack? This is unfortunate and I can't see a way around it as well. so yeah. makes sense

Reply:
Exactly, it's basically imperatively pushed routes that necessitates the use of RouteMatchList.